sync/pending_polling.go

```diff
@@ -0,0 +1,672 @@
+				return true // switch to pre_confirmed
+			}
+			if err := s.StorePending(p); err != nil {
+				s.log.Debugw("Error storing pending block", "err", err)
+			}
+		}
+	}
+}
+
+// handleHeadInPreConfirmed processes a new head during the pre_confirmed phase.
+// It computes nextHeight = head.Number + 1, clears any staged pre_latest if the
+// head catches up to the current target, and stores an empty pre_confirmed when
+// advancing. Returns updated (targetNum, stagedPreLatest).
+func (s *Synchronizer) handleHeadInPreConfirmed(
+	head *core.Block,
+	targetNum uint64,
+	stagedPreLatest *core.PreLatest,
+) (uint64, *core.PreLatest) {
+	next := head.Number + 1
+	if next == targetNum && stagedPreLatest != nil {
+		s.UpdatePreLatestAttachment(targetNum, nil)
+		return targetNum, nil
+	}
+
+	if next > targetNum {
+		s.targetPreConfirmedNum.Store(next)
+		if err := s.storeEmptyPreConfirmed(head.Header, nil); err != nil {
+			s.log.Debugw("Error storing empty pre_confirmed (from head)", "err", err)
+		}
+		return next, nil
+	}
+
+	return targetNum, stagedPreLatest
+}
+
+// handlePreLatest processes an incoming pre_latest during the pre_confirmed phase.
+// If it raises the target, it stages the attachment and stores a baseline with it.
+// Returns updated (targetNum, stagedPreLatest).
+func (s *Synchronizer) handlePreLatest(
+	pl *core.PreLatest,
+	targetNum uint64,
+	stagedPreLatest *core.PreLatest,
+) (uint64, *core.PreLatest) {
+	next := pl.Block.Number + 1
+
+	if next > targetNum {
+		s.targetPreConfirmedNum.Store(next)
+		if err := s.storeEmptyPreConfirmed(pl.Block.Header, pl); err != nil {
+			s.log.Debugw("Error storing empty pre_confirmed (with pre_latest)", "err", err)
+		}
+		return next, pl
+	}
+
+	return targetNum, stagedPreLatest
+}
+
+// handlePreConfirmed finalises when the polled pre_confirmed equals the target.
+// It attaches the staged pre_latest (if still valid), stores it, and feeds if changed.
+func (s *Synchronizer) handlePreConfirmed(
+	pc *core.PreConfirmed,
+	targetNum uint64,
+	stagedPreLatest *core.PreLatest,
+) {
+	if pc.Block.Number != targetNum {
+		return
+	}
+
+	pc.WithPreLatest(stagedPreLatest)
+	changed, err := s.StorePreConfirmed(pc)
+	if err != nil {
+		s.log.Debugw("Error while trying to store pre_confirmed block", "err", err)
+		return
+	}
+	if changed {
+		s.pendingDataFeed.Send(pc)
+	}
+}
+
+// runPreConfirmedPhase coordinates baselines, pre_latest attachments, and final storage.
+// Key rule: if head for target-1 arrives (i.e., head.Number+1 == targetNum) after staging a pre_latest,
+// we clear the staged attachment and also clear the stored baseline’s attachment (if present).
+func (s *Synchronizer) runPreConfirmedPhase(ctx context.Context, headsSub *feed.Subscription[*core.Block]) {
+	preLatestCh := make(chan *core.PreLatest, 1)
+	preConfirmedCh := make(chan *core.PreConfirmed, 1)
+
+	go s.pollPreLatest(ctx, preLatestCh)
+	go s.pollPreConfirmed(ctx, preConfirmedCh)
+
+	var (
+		targetNum       uint64
```

Comment:
I think `targetNum` is a confusing name, whether it's target of head, prelatest or preconfirmed.